Add drop_remainder argument to bucket_by_sequence_length

yweweler opened the pull request:
`tf.data.experimental.bucket_by_sequence_length` does not allow to drop
the last batch in case it has fewer than `batch_size` elements.
This patch does implement `drop_remainder` for `bucket_by_sequence_length`
to enable thhis behaviour.

`drop_remainder` is optinal and set to `False` by default to maintain
compatibility.

I will add a PR to update the documentation of bucket_by_sequence_length once this is merged.

yweweler committed:
Add drop_remainder argument to bucket_by_sequence_length

`tf.data.experimental.bucket_by_sequence_length` does not allow to drop
the last batch in case it has fewer than `batch_size` elements.
This patch does implement `drop_remainder` for `bucket_by_sequence_length`
to enable thhis behaviour.

`drop_remainder` is optinal and set to `False` by default to maintain
compatibility.

mrry reviewed (changes requested):
Can you please add a test for the new arguments?

yweweler committed:
Add drop_reminder support to the test of bucketing of sparse tensors

yweweler committed:
Add a drop_reminder version of the testBucket test

Squeezing both cases into a single test would make it way to
complicated. Therefore, I created a separate test.

yweweler committed:
Fix pylint warnings

* Missing argument drop_remainder in the testBucketSparse test.
* Insert line breaks

yweweler commented:
@mrry , I added `drop_remainder` tests to the existing test cases.
Are these sufficient?

I run the tests using the nightly-docker file (CPU only).

mrry commented on tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py:
```diff
@@ -306,16 +422,18 @@ def _compute_batches(dataset):
+          no_padding=True,
+          drop_remainder=drop_remainder))
+      batches = _compute_batches(dataset)
+      expected_batches = _compute_expected_batches()
```
Should this call take a `drop_remainder` argument?

yweweler replied:
Yap, sorry about that, it slipped into the pylint commit 6842678

yweweler committed:
Fix copy paste bug

Pasted the wrong variable names from the docker test image.

yweweler commented on tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py:
```diff
@@ -154,9 +154,9 @@ def _test_bucket_by_padding(no_padding):
+        generated_sums[length] = 0
         # Initialize the individual batch sizes.
-        expected_batch_sizes[length] = []
+        generated_batch_sizes[length] = []
```
I initialized the wrong variables due to a copy paste failure.

yweweler commented:
@mrry  The `bucket_by_sequence_length` test passes with the latest commit 4ddd2ab.

The only other test that is failing is the `api_compatibility_test`.
However, I am not sure how these work. Am I responsible for taking action when they fail?

mrry commented:
@yweweler I think you can update the necessary API files by running the following two commands:

```
bazel build //tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens=True
```

yweweler commented:
> @yweweler I think you can update the necessary API files by running the following two commands:
> 
> ```
> bazel build //tensorflow/tools/api/tests:api_compatibility_test
> bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens=True
> ```

@mrry I have to admit that I am a bit stuck.
I am using the two lines from [tensorflow/tools/api/tests/README.txt](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/tests/README.txt) to build and update the goldens.
```
$ bazel build //tensorflow/tools/api/tests:api_compatibility_test
$ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
```
The build part works flawlessly.
I assured that all dependencies are installed properly and that building tensorflow as a whole also works.
However, updating the goldens using `api_compatibility_test` crashes.

```
(tensorflow_27_venv) user:~/downloads/tensorflow$ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
Traceback (most recent call last):
  File "/home/user/downloads/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 36, in <module>
    from tensorflow._api.v2 import v2 as tf_v2
  File "/home/user/downloads/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/_api/v2/v2.py", line 325, in <module>
    if _tf_api_dir not in __path__:
NameError: name '__path__' is not defined
```

There are some issues in the repo suggesting that the procedure can only be executed using python2.7.
However, I tried doing the whole procedure both using python2.7 and python3.5 and ran into the same error every time.

I also tried the same thing on older commits from the `master` branch. Sadly with the same outcome.

yweweler commented:
> NameError: name '\_\_path__' is not defined

I already fiddled around with the "\_\_path__"  variable by manually initializing it inside `/home/user/downloads/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/_api/v2/__init__.py`

However, this just leads to more errors, like:
```
ImportError: cannot import name v2
```

mrry commented:
@annarev Is thie `api_compatibility_test` currently expected to work in open source?

yweweler commented:
> @annarev Is thie `api_compatibility_test` currently expected to work in open source?

I would like to prevent this to fall into decline.

Guessing from the `api_compatibility_test` code I think that the changes to the goldens are minimal.
Therefore I could patch the goldens manually and see if the test passes. I don't really like working around this but currently I see no other option.

Either way it would be great to have some clarification on the state of `api_compatibility_test`.

yweweler committed:
Update tensorflow.data.experimental goldens for v1

Add the drop_remainder argument for bucket_by_sequence_length to the goldens.

yweweler committed:
Update tensorflow.data.experimental goldens for v2

Add the drop_remainder argument for bucket_by_sequence_length to the goldens.

yweweler commented:
@mrry I patched the goldens for both v1 and v2 (Not sure if this is desired).

I also retried executing the test with different docker images (`devel-py3` and `nightly-devel`), sadly with the  same outcome as described above.

As I can't test the changes using `api_compatibility_test`, it would be great if we could run the through kokoro.

harshini-gadige commented:
@jsimsa -  Could you PTAL at this ?

jsimsa commented on tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py:
```diff
@@ -71,6 +71,130 @@ def _get_record_shape(sparse):
+                       .format(sorted(expected_lengths),
+                               sorted(generated_lengths)))
+
+    for no_padding in (True, False):
```
use `@parameterized` for this (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/kernel_tests/map_and_batch_test.py for an example)

jsimsa reviewed (changes requested):
did a pass and left some minor comments

yweweler committed:
Fix typos

yweweler committed:
Rewrite grouping tests to be @parameterized tests

yweweler commented:
@jsimsa I made the changes you requested.

yweweler committed:
Merge branch 'master' into yweweler-bucketing-add-drop-remainder

jsimsa commented:
@shivaniag could you please review this PR, the newly introduced test still uses session and I believe it should be switched to the new style which you introduced

yweweler commented:
> @shivaniag could you please review this PR, the newly introduced test still uses session and I believe it should be switched to the new style which you introduced

Thats a good point. I did not notice that @shivaniag updated the tests in the meantime.

I was going through his commit 4caa17185fb39420c87373af77b538e9016be46d trying to understand the changes he made to the other tests.
Before updating the PR to match his changes I need to know why the following change is needed.

https://github.com/tensorflow/tensorflow/blob/4caa17185fb39420c87373af77b538e9016be46d/tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py#L128-L130

@shivaniag Why do we need to disable the sum_check when not executing eagerly?

shivaniag commented:
@yweweler: No, we don't need to disable sum_check when executing eagerly, there is a need to add eager coverage for this test which has been left as a TODO in the last commit.

shivaniag commented on tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py:
```diff
@@ -72,10 +74,141 @@ def _get_record_shape(sparse):
+    ("WithoutPadding", True),
+    ("WithPadding", False),
+  )
+  def testBucketDropReminder(self, param_no_padding):
```
make this testSkipEagerBucketDropReminder for now and add a TODO note for eager coverage.

yweweler committed:
Update testBucketDropReminder test case

See #24071

yweweler commented:
@shivaniag  This should implement the changes the requested changes.

shivaniag reviewed (dismissed):
Approving the test changes.

yweweler committed:
Fix pylint warnings from Ubuntu Sanity test

yweweler commented:
Not sure why I can't see the test results from the kokoro run any more.
However, apart from me having forgot to run pylint on the last commit, the other errors were not related to this PR (as far as I can tell.)
